Undo Phase 1

halgari opened the pull request:
Adds phase 1 of Undo for loadouts. 

* Implements CLI commands for showing revisionss `loadout revisions`
* Implements a CLI command for reverting a loadout to a revision `loadout revert`
* Uses Cascade, and has a (not fully generic yet) implementation of a "Redo" method for MnemonicDB

None of this is yet hooked into the UI, but there is a test that does basic validaton of the results and queries.

Example usage, currently reverting is a bit slow, as is the next boot-up after a revert (thanks to the TreeAnalyzer), but that will all be improved in future PRs.

![alacritty_tXTH2zm0UZ](https://github.com/user-attachments/assets/593e1892-5479-435b-8bfa-58c593ba188c)

erri120 commented:
I ran the `loadout revert` CLI command a couple of times and managed to get this exception somehow:

```
An error occurred while executing the commandError: RocksDbSharp.RocksDbException: Corruption: Compaction sees out-of-order keys.
   at RocksDbSharp.Native.rocksdb_write(IntPtr db, IntPtr options, IntPtr batch)
   at RocksDbSharp.RocksDb.Write(WriteBatch writeBatch, WriteOptions writeOptions)
   at NexusMods.MnemonicDB.Storage.RocksDbBackend.Batch.Commit()
   at NexusMods.MnemonicDB.Storage.DatomStore.LogDatoms[TSource](TSource datoms, Boolean advanceTx, Boolean
enableStats)
   at NexusMods.MnemonicDB.InternalTxFunctions.IndexSegmentTransaction.Execute(DatomStore store)
   at NexusMods.MnemonicDB.Storage.DatomStore.Log(IInternalTxFunctionImpl pendingTransaction)
   at NexusMods.MnemonicDB.Storage.DatomStore.ConsumeTransactions()
--- End of stack trace from previous location ---
   at NexusMods.MnemonicDB.Storage.DatomStore.TransactAsync(IInternalTxFunction fn)
   at NexusMods.MnemonicDB.Connection.Transact(IInternalTxFunction fn)
   at NexusMods.MnemonicDB.Transaction.Commit()
   at NexusMods.DataModel.Undo.UndoService.RevertTo(LoadoutRevision revisionRevision) in
/mnt/redline/projects/NexusMods/app/code-review/src/NexusMods.DataModel/Undo/UndoService.cs:line 120
   at NexusMods.DataModel.CommandLine.Verbs.LoadoutManagementVerbs.Revert(IRenderer renderer, ReadOnly
loadout, Int32 revisionNumber, IConnection connection, UndoService undoService, CancellationToken token) in
/mnt/redline/projects/NexusMods/app/code-review/src/NexusMods.DataModel/CommandLine/Verbs/LoadoutManagement
Verbs.cs:line 275
   at NexusMods.SingleProcess.CommandHandler.InvokeAsync(InvocationContext context) in
/mnt/redline/projects/NexusMods/app/code-review/src/NexusMods.SingleProcess/CommandHandler.cs:line 38
```

halgari commented:
ooh, that's not good, I'll take a look

halgari commented:
Added a test, and was able to trigger the error. I had optimized MnemonicDB's sorting of case insensitive strings, and something had shifted in the ordering. Previously `A` was sorted after `_`, and after the change this order was inverted. I fixed the comparison code and it now works as expected.
